File: compression.py
t = int(input())
out_sequence = []
# sequence = []
for _ in range(t):
    res = []
    n = int(input())
    sequence = list(map(int, input().split()))
    sequence.append(sequence[-1] + 10)  # Fake number
    if len(sequence) == 2:
        out_sequence.append([(sequence[0], 0)])
        continue
    i = 1
    while i < len(sequence):
        step = 0
        count = 0
        start = sequence[i - 1]
        if sequence[i] == sequence[i - 1] + 1:
            step = 1
        elif sequence[i] == sequence[i - 1] - 1:
            step = -1
        while step != 0 and sequence[i] == sequence[i - 1] + step and i < len(sequence):
            count += 1
            i += 1
        res.append((start, count * step))
        i += 1
    out_sequence.append(res)

for test in range(t):
    print(len(out_sequence[test])*2)
    for dig in out_sequence[test]:
        print(dig[0], end=" ")
        print(dig[1], end=" ")
    print()

File: compression.go
package main

import "fmt"

func main() {
	var n int
	fmt.Scan(&n)
	outSequence := make([][]int, n)

	for j := 0; j < n; j++ {
		var len int
		fmt.Scan(&len)
		sequence := make([]int, len)
		for i := 0; i < len; i++ {
			fmt.Scan(&sequence[i])
		}
		sequence = append(sequence, sequence[len-1]+10) // Fake number
		if len == 1 {
			outSequence[j] = append(outSequence[j], sequence[0])
			outSequence[j] = append(outSequence[j], 0)
			continue
		}
		for i := 1; i < len+1; i++ {
			step := 0
			count := 0
			start := sequence[i-1]
			if sequence[i] == sequence[i-1]+1 {
				step = +1
			} else if sequence[i] == sequence[i-1]-1 {
				step = -1
			}
			for step != 0 && sequence[i] == sequence[i-1]+step && i < len {
				count++
				i++
			}
			outSequence[j] = append(outSequence[j], start)
			outSequence[j] = append(outSequence[j], count*step)
		}
	}
	for j := 0; j < n; j++ {
		fmt.Println(len(outSequence[j]))
		for _, dig := range outSequence[j] {
			fmt.Printf("%d ", dig)
		}
		fmt.Println()
	}
}
